Java basic lazy method conversion

smowton opened the pull request:
This patch series adds lazy method conversion for Java, such that static methods need a direct caller and virtual methods need a constructor site and a potential virtual callsite before they get converted. This can significantly shrink the amount of code converted when large libraries are in play.

This will need rebasing once some of the existing Java improvements in test-gen are merged in; meanwhile the true diff can be seen at https://github.com/diffblue/cbmc-testgen/compare/cristina-david/test_gen...smowton:java_basic_lazy_conversion?expand=1

Before this becomes mergable I'll also make lazy method loading optional based on a command-line parameter.

kroening commented:
May I ask for some sensible squashing?

smowton commented:
@kroening were you looking at the real diff (https://github.com/diffblue/cbmc-testgen/compare/cristina-david/test_gen...smowton:java_basic_lazy_conversion?expand=1) ? If even that is too many commits then I can squash them as you like.

smowton commented:
Just rebased this. Untested against master so stuff may be broken, but it's now optional and at least compiles. TODO tomorrow: testing, check whether the rebase has introduced anything transparently silly into the commit history.

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -179,6 +179,66 @@ const exprt java_bytecode_convert_methodt::variable(
+
+void java_bytecode_convert_method_lazy(
+  const symbolt& class_symbol,
+  const irep_idt method_identifier,
```
why not a const ref?

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -179,6 +179,66 @@ const exprt java_bytecode_convert_methodt::variable(
+
+  if(method_symbol.base_name=="<init>")
+  {
+    method_symbol.pretty_name=id2string(class_symbol.pretty_name)+"."+
```
break after =

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -179,6 +179,66 @@ const exprt java_bytecode_convert_methodt::variable(
+    member_type.set(ID_constructor, true);
+  }
+  else
+    method_symbol.pretty_name=id2string(class_symbol.pretty_name)+"."+
```
break after =

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -193,14 +253,13 @@ void java_bytecode_convert_methodt::convert(
     id2string(class_symbol.name)+"."+id2string(m.name)+":"+m.signature;
   method_id=method_identifier;
 
+  const auto& old_sym=symbol_table.lookup(method_identifier);
```
&old_sym

peterschrammel commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -170,6 +181,287 @@ bool java_bytecode_languaget::parse(
+\*******************************************************************/
+
+static irep_idt get_virtual_method_target(
+  const std::set<irep_idt>& needed_classes,
```
cpplnt

peterschrammel commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -170,6 +181,287 @@ bool java_bytecode_languaget::parse(
+
+\*******************************************************************/
+
+static void initialise_needed_classes(
```
initialize

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2140,13 +2212,17 @@ void java_bytecode_convert_method(
   bool disable_runtime_checks,
-  size_t max_array_length)
+  size_t max_array_length,
+  std::vector<irep_idt> *needed_methods,
```
why not a reference?

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1033,7 +1092,11 @@ codet java_bytecode_convert_methodt::convert_instructions(
             if(as_string(arg0.get(ID_identifier))
                .find("<init>")!=std::string::npos)
+            {
+              if(needed_classes)
```
Using a flag would avoid pointers.

peterschrammel commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1115,11 +1178,16 @@ codet java_bytecode_convert_methodt::convert_instructions(
         // static binding
         call.function()=symbol_exprt(arg0.get(ID_identifier), arg0.type());
+        needed_methods->push_back(arg0.get(ID_identifier));
+        if(needed_methods)
```
???

peterschrammel commented on src/util/language_file.cpp:
```diff
@@ -64,6 +64,13 @@ void language_filet::get_modules()
 
+void language_filet::convert_lazy_method(
+  const irep_idt &id,
+  symbol_tablet &symtab)
```
symbol_table preferred

peterschrammel commented on src/util/language_file.h:
```diff
@@ -42,6 +42,10 @@ class language_filet
 
+  void convert_lazy_method(
+    const irep_idt &id,
+    symbol_tablet &symtab);
```
symbol_table preferred

peterschrammel commented on src/util/language_file.h:
```diff
@@ -57,9 +61,15 @@ class language_filest:public messaget
+  // Contains pointers into filemapt!
+  // This is safe-ish as long as this is std::map.
+  typedef std::map<irep_idt, language_filet*> lazymethodmapt;
+  lazymethodmapt lazymethodmap;
```
consider using _ to_separate_words

peterschrammel commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -27,11 +28,15 @@ class java_bytecode_convert_classt:public messaget
     bool _disable_runtime_checks,
-    size_t _max_array_length):
+    size_t _max_array_length,
+    lazy_methodst& _lm,
```
_lazy_methods preferred

peterschrammel commented on src/util/language_file.h:
```diff
@@ -75,10 +85,23 @@ class language_filest:public messaget
+    const irep_idt &id,
+    symbol_tablet &symtab)
+  {
+    return lazymethodmap.at(id)->convert_lazy_method(id, symtab);
```
Are we sure id is in the map?

smowton replied:
Yes, it is always an error to try to do second-stage conversion on a method that hasn't undergone first stage. Added a comment.

smowton commented:
All comments applied, except as noted

smowton commented:
Re: pointers, to me it seems like more of a mess to pass a reference to an empty map and a bool indicating whether to use it or not, compared to a pointer. How strongly do you feel on that? How about some really basic `boost::optional`-alike instead?

peterschrammel commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -212,6 +637,26 @@ bool java_bytecode_languaget::typecheck(
   return false;
 }
 
+void java_bytecode_languaget::lazy_methods_provided(std::set<irep_idt> &methods) const
```
function comment block missing (several instances)

peterschrammel commented on src/java_bytecode/java_bytecode_language.h:
```diff
@@ -16,6 +16,20 @@ Author: Daniel Kroening, kroening@kroening.com
+};
+
+typedef std::pair<
+          const symbolt*,
```
`symbolt *` (several instances)

peterschrammel commented on src/java_bytecode/java_bytecode_language.h:
```diff
@@ -69,9 +83,14 @@ class java_bytecode_languaget:public languaget
+  virtual void convert_lazy_method(const irep_idt &id, symbol_tablet &);
 
 protected:
+  bool do_ci_lazy_method_conversion(symbol_tablet &, lazy_methodst &);
```
maybe add a comment on what the highly overloaded abbreviation CI means in this context

peterschrammel commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -200,10 +492,143 @@ bool java_bytecode_languaget::typecheck(
+          debug() << "Skip " << mname << eom;
+          continue;
+        }
+        debug() << "CI lazy methods: elaborate " << mname << eom;
```
CI...

peterschrammel commented:
Yes, the clean way is to use an option pattern. I'm fine with either solution for it (except requiring boost for doing that). It's just to make sure that a programming error asserts rather than segfaults.

smowton commented:
Just updated, taking all comments into account. Introduced a simple `safe_pointer` type that provides explicit null checks and throws if `get()` is called against a null pointer, necessarily requiring a check with `operator bool`. Similarly creating one of these requires explicit permission to admit a null argument.

Also fixed a couple of bugs I noticed that were causing `cbmc-java` test failures; these have been squashed into existing commits to avoid bisection problems in future.

peterschrammel commented on src/util/safe_pointer.h:
```diff
@@ -0,0 +1,56 @@
+/*******************************************************************\
```
It would be good if this was interface-compatible with unique_ptr, etc, e.g. use constructors and overload * and -> for convenient access

peterschrammel commented on src/util/safe_pointer.h:
```diff
@@ -0,0 +1,56 @@
+
+  T *get()
+  {
+    if(!(*this))
```
I'd prefer an assertion here, because it's clearly a programming error and not something that someone should be able to catch.

peterschrammel commented on src/util/safe_pointer.h:
```diff
@@ -0,0 +1,56 @@
+    T *target)
+  {
+    if(!target)
+      throw "initialised safe pointer with null";
```
initialized (sorry!)

peterschrammel commented on src/util/language_file.h:
```diff
@@ -54,15 +58,21 @@ class language_filet
-  modulemapt modulemap;
+  // Contains pointers into filemapt!
+  // This is safe-ish as long as this is std::map.
+  typedef std::map<irep_idt, language_filet*> lazy_method_mapt;
```
language_filet *

peterschrammel commented on src/java_bytecode/java_entry_point.h:
```diff
@@ -28,6 +28,7 @@ typedef struct
   const irep_idt &main_class,
-  message_handlert &);
+  message_handlert &,
+  bool allow_no_body = false);
```
no spaces

peterschrammel commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -170,6 +181,287 @@ bool java_bytecode_languaget::parse(
+
+static void get_virtual_method_targets(
+  const code_function_callt &c,
+  const std::set<irep_idt>& needed_classes,
```
check & in whole PR

peterschrammel commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -170,6 +181,287 @@ bool java_bytecode_languaget::parse(
+      else
+      {
+        const auto &entry=findit->second;
+        if(entry.parents.size()==0)
```
.empty()

smowton commented:
All changes applied, and in particular `safe_pointer` gains operator*, operator->.

smowton commented:
This nearly works, but is producing a baffling test failure under Linux/Clang. Waiting for a rebuild with a more verbose version of the failed-test printer Perl script.

smowton commented:
Just re-pushed after a rebase to re-test (with test regex dialect changed).

I think the failed-tests improvement is worth keeping, though I can split that off as an individual PR if you like.

peterschrammel commented:
@kroening, this is also ready now.

kroening commented:
What is the purpose of the --lazy-methods parameter?
Is there a scenario in which it is not sensible to lazily load?

smowton commented:
If the method for spotting constructors misfires and therefore underestimates the possible virtual callsite targets, e.g. if a potential callee was introduced via reflection. It also provides a way for a potentially-buggy new feature to be eased in with a possibility to revert the behaviour easily.

kroening commented:
Ok, I'll merge for now, but the goal should be to get it so robust that the parameter is just not needed.
